### PR TITLE
Fix #31191: duplicated "scope" parameter

### LINKED ISSUE #31191: [BUG] The parameter 'scope' is duplicated
v.4.4.0-beta.1

**Describe the bug**
[invalid_request] AADSTS9000411: The request is not properly formatted. The parameter 'scope' is duplicated.
AadOAuth2AuthorizationCodeGrantRequestEntityConverter will be executed multiple times and adds parameter "scope" into request body.
1.Attempt:
openid profile offline_access https://graph.microsoft.com/Directory.Read.All

2.Attempt:
openid profile offline_access https://graph.microsoft.com/Directory.Read.All
openid profile offline_access https://graph.microsoft.com/Directory.Read.All

3.Attempt:
openid profile offline_access https://graph.microsoft.com/Directory.Read.All
openid profile offline_access https://graph.microsoft.com/Directory.Read.All
openid profile offline_access https://graph.microsoft.com/Directory.Read.All

...

**To Reproduce**
Added proxy support. Used RestTemplateCustomizer:
```java
@Bean
public RestTemplateCustomizer proxyCustomizer() {
  return restTemplate -> {
    Proxy proxy = new Proxy(Type.HTTP, new InetSocketAddress("127.0.0.1", 8888));
    SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
    requestFactory.setProxy(proxy);
    restTemplate.setRequestFactory(requestFactory);
  };
}
```
Used the default AadWebSecurityConfigurerAdapter for web security config.
Authorization/Authentification fails: org.springframework.security.oauth2.client.endpoint.DefaultAuthorizationCodeTokenResponseClient#getTokenResponse 
-> access token in OAuth2AccessTokenResponse is empty.

***Exception or Stack Trace***
```log
Oauth2 flow:
[http-nio-8080-exec-5]: org.springframework.security.web.FilterChainProxy: Securing GET /oauth2/authorization/azure
[http-nio-8080-exec-5]: org.springframework.security.web.context.SecurityContextPersistenceFilter: Set SecurityContextHolder to empty SecurityContext
[http-nio-8080-exec-5]: org.springframework.security.web.DefaultRedirectStrategy: Redirecting to https://login.microsoftonline.com/xyz/oauth2/v2.0/authorize?response_type=code&client_id=xyzxyz&scope=openid%20profile%20offline_access%20https://graph.microsoft.com/Directory.Read.All&state=xyzard%3D&redirect_uri=http://localhost:8080/nautilus/newbusiness/login/oauth2/code/&nonce=xyzafg

[http-nio-8080-exec-6]: org.springframework.security.web.FilterChainProxy: Securing GET /login/oauth2/code/?code=0.ATAxyzae5c

tokenResponse.getAccessToken()  is null and fails at 
`if (CollectionUtils.isEmpty(tokenResponse.getAccessToken().getScopes())) {`

`java.lang.NullPointerException: null
	at org.springframework.security.oauth2.client.endpoint.DefaultAuthorizationCodeTokenResponseClient.getTokenResponse(DefaultAuthorizationCodeTokenResponseClient.java:80)
	at org.springframework.security.oauth2.client.endpoint.DefaultAuthorizationCodeTokenResponseClient.getTokenResponse(DefaultAuthorizationCodeTokenResponseClient.java:57)
	at org.springframework.security.oauth2.client.oidc.authentication.OidcAuthorizationCodeAuthenticationProvider.getResponse(OidcAuthorizationCodeAuthenticationProvider.java:170)
	at org.springframework.security.oauth2.client.oidc.authentication.OidcAuthorizationCodeAuthenticationProvider.authenticate(OidcAuthorizationCodeAuthenticationProvider.java:144)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:182)`
```

**Screenshots**
After 8 attempts:
image

**Setup (please complete the following information):**
 - OS: Windows
 - IDE: IntelliJ
 - Library/Libraries: Azure SDK 4.4.0-beta.1 from SEP-28-2022 main-branch.
 - Java version: 11
 - App Server/Environment: Spring Boot Oauth2 (tomcat)+Azure
 - Frameworks: Spring Boot 2.6.7

### PR DESCRIPTION
**Pull request by chenrujun**
Fix #31191. duplicated "scope" parameter.

### Root Cause Analysis

This bug is caused by https://github.com/Azure/azure-sdk-for-java/pull/30456/commits/a79493d3c8d54117a89c8135d8c079075474c244

image

[AbstractOAuth2AuthorizationGrantRequestEntityConverter#addParametersConverter](https://github.com/spring-projects/spring-security/blob/5.7.3/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AuthorizationGrantRequestEntityConverter.java#L156) is not idempotent, `AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter#convert` execute multiple times will cause duplicated parameters.

### How To Fix
Move the following code into constructor to make sure they only execute once.
```java
addHeadersConverter(this::getHttpHeaders);
addParametersConverter(this::getHttpBody);
```

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java [hunk 1]
```diff
@@ -28,22 +36,15 @@
     protected abstract String getApplicationId();
 
     @Override
-    @SuppressWarnings("unchecked")
     public RequestEntity<?> convert(OAuth2AuthorizationCodeGrantRequest request) {
-        addHeadersConverter(headersConverter);
-        addParametersConverter(parametersConverter);
         return super.convert(request);
     }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:41]: should we remove this now?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java [hunk 2]
```diff
@@ -28,22 +36,15 @@
...
-    private final Converter<OAuth2AuthorizationCodeGrantRequest, HttpHeaders> headersConverter = (request) -> getHttpHeaders();
-
-    private final Converter<OAuth2AuthorizationCodeGrantRequest, MultiValueMap<String, String>> parametersConverter = this::getHttpBody;
-
     /**
      * Additional default headers information.
      * @return HttpHeaders
      */
-    public HttpHeaders getHttpHeaders() {
+    public HttpHeaders getHttpHeaders(OAuth2AuthorizationCodeGrantRequest request) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:47]: Why changing this?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java (hunk 1 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java]: Updated.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java (hunk 2 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:41]: Just accept suggestion from here: https://github.com/Azure/azure-sdk-for-java/pull/31213/files#r990644464
image

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:41]: Should `getHttpHeaders` and `getHttpBody` be `protected` then?

**Review by saragluna — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java [hunk 3]
```diff
@@ -20,30 +19,26 @@
 public abstract class AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter
     extends OAuth2AuthorizationCodeGrantRequestEntityConverter {
 
+    private static final MultiValueMap<String, String> EMPTY_MULTI_VALUE_MAP =
+            new MultiValueMapAdapter<>(Collections.emptyMap());
+
+    protected AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter() {
+        addHeadersConverter(this::getHttpHeaders);
+        addParametersConverter(this::getHttpBody);
+    }
```
Comment (moarychan) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:28]: We should explicitly use this default constructor in the subclasses  
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/webapp/AadOAuth2AuthorizationCodeGrantRequestEntityConverter.java#L34-L36

Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/oauth2/AbstractOAuth2AuthorizationCodeGrantRequestEntityConverter.java:28]: Not necessary.
Refs:
https://stackoverflow.com/questions/2054022/is-it-unnecessary-to-put-super-in-constructor

Test:

image

**Pull request merged**